### PR TITLE
Added resolution display in options and added a tutorial highlight

### LINKED ISSUE #2857: Show the current rendering resolution in options
It's not really possible in Godot to not render at native resolution, but we should communicate better that that is what we do.
For example in this video there's an user who has had bad experiences with automatic resolution (which we'll need to have as we can't really change that in Godot): https://www.youtube.com/watch?v=WfIO6HRXkH8
and I think that experience would be improved if we showed the current resolution next to the "auto", for example like `auto (1920x1080)` should be relatively easy to grab the resolution from Godot (and also subscribe the options menu to resolution change events from Godot).

### LINKED ISSUE #3103: Modify editor redo button tutorial to use the GUI control highlight
so that the user is very obviously pointed to the redo button and they can't skip the tutorial

### PR DESCRIPTION
**Pull request by harchrisdytian**
**Brief Description of What This PR Does**

this PR changes it so that when the redo tutorial displays it displays a control syntax display over the button.
it also modifies the settings screen to display the current resolution the game is at while it is in both fullscreen mode and in regular

**Related Issues**

Closes #3103
Closes #2857

**Progress Checklist**

Note: before starting this checklist the PR should be marked as non-draft.

- [x] PR author has checked that this PR works as intended and doesn't
      break existing features:
      https://wiki.revolutionarygamesstudio.com/wiki/Testing_Checklist
      (this is important as to not waste the time of Thrive team
      members reviewing this PR)
- [x] Initial code review passed (this and further items should not be checked by the PR author)
- [x] Functionality is confirmed working by another person (see above checklist link)
- [x] Final code review is passed and code conforms to the 
      [styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md).

Before merging all CI jobs should finish on this PR without errors, if
there are automatically detected style issues they should be fixed by
the PR author. Merging must follow our
[styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md#git).

### CODE REVIEW

File: src/general/OptionsMenu.cs [hunk 1]
```diff
@@ -420,6 +425,9 @@ public override void _Ready()
         guiLightEffectsToggle.RegisterToolTipForControl("guiLightEffects", "options");
         assumeHyperthreading.RegisterToolTipForControl("assumeHyperthreading", "options");
         unsavedProgressWarningEnabled.RegisterToolTipForControl("unsavedProgressWarning", "options");
+
+        // Makes it so that when the viewport changes size it automatically displays
+        GetViewport().Connect("size_changed", this, "DisplayResolution");
```
Comment (hhyyrylainen) [inline src/general/OptionsMenu.cs:430]: I think `_Notification` might be able to already receive the size changed notification without this signal connection required.

File: src/general/OptionsMenu.cs [hunk 2]
```diff
@@ -577,29 +586,41 @@ private void SwitchMode(OptionsMode mode)
         switch (mode)
         {
             case OptionsMode.MainMenu:
-            {
-                tutorialsEnabled.Hide();
-                optionsMode = OptionsMode.MainMenu;
-                break;
-            }
+                {
```
Comment (hhyyrylainen) [inline src/general/OptionsMenu.cs:589]: Here's the broken indents that jetbrains is complaining about.

File: src/general/OptionsMenu.cs [hunk 3]
```diff
@@ -577,29 +586,41 @@ private void SwitchMode(OptionsMode mode)
...
+                    optionsMode = OptionsMode.InGame;
+                    break;
+                }
 
             default:
                 throw new ArgumentException("Options menu SwitchMode called with an invalid mode argument");
         }
     }
 
+    // Displays the current Resolution in options.
```
Comment (hhyyrylainen) [inline src/general/OptionsMenu.cs:612]: Method documentation comments need to use the xml format.

File: src/general/OptionsMenu.cs [hunk 4]
```diff
@@ -577,29 +586,41 @@ private void SwitchMode(OptionsMode mode)
...
+    private void DisplayResolution()
+    {
+        if (resolution != null)
+        {
+            System.Threading.Thread.CurrentThread.CurrentCulture =
```
Comment (hhyyrylainen) [inline src/general/OptionsMenu.cs:617]: Don't randomly change the language...

File: src/general/OptionsMenu.cs [hunk 5]
```diff
@@ -577,29 +586,41 @@ private void SwitchMode(OptionsMode mode)
...
+                System.Globalization.CultureInfo.CreateSpecificCulture("en-UK");
+            resolution.Text = "AUTO" + "(" + GetViewportRect().Size.x.ToString()
```
Comment (hhyyrylainen) [inline src/general/OptionsMenu.cs:619]: This needs to use the translation system. I suggest key name `AUTO_RESOLUTION` with the English content being `Auto ({0}x{1})` and then formatting the width and height to that string.

File: src/general/OptionsMenu.cs (hunk 1 above)
Comment (hhyyrylainen) [reply src/general/OptionsMenu.cs:430]: Also you need to call the update resolution in _Notification when language changes.

File: src/microbe_stage/editor/MicrobeEditorReportComponent.tscn [hunk 6]
```diff
@@ -73,9 +73,6 @@ anchor_bottom = 1.0
 margin_top = -0.200001
 theme = ExtResource( 9 )
 script = ExtResource( 11 )
-__meta__ = {
```
Comment (hhyyrylainen) [inline src/microbe_stage/editor/MicrobeEditorReportComponent.tscn:76]: Random changes to unrelated files that aren't related to the PR at all should be reverted.

**Comment by hhyyrylainen** [comment]
I didn't test yet but this looks like it would work. I've requested quite few bug and code style fixes, though. Also in general you should put only one "feature" in a PR so this should have been 2 PRs: one for the options and one for the tutorial. That would also help with making a better name for the PR. And one last thing: please request the programming team for review instead of random people, most of who are inactive.

File: src/microbe_stage/editor/MicrobeEditorWorld.tscn [hunk 7]
```diff
@@ -27,12 +25,7 @@ shader_param/maskTexture = ExtResource( 3 )
 
 [node name="MicrobeEditorWorld" type="Node"]
 
-[node name="PrimaryCamera" type="Camera" parent="." instance=ExtResource( 5 )]
```
Comment (hhyyrylainen) [inline src/microbe_stage/editor/MicrobeEditorWorld.tscn:30]: These look very much incorrect changes that should not be in this PR, please remove these, otherwise this can't be merged without breaking existing features.

File: src/microbe_stage/editor/CellEditorComponent.GUI.cs [hunk 8]
```diff
@@ -24,6 +24,13 @@ public void SendUndoToTutorial(TutorialState tutorial)
         tutorial.AutoEvoPrediction.EditorAutoEvoPredictionPanel = autoEvoPredictionPanel;
     }
 
+    public void SendRedoToTutorial(TutorialState tutorial)
```
Comment (hhyyrylainen) [inline src/microbe_stage/editor/CellEditorComponent.GUI.cs:27]: Maybe it would make more sense to combine this method with `SendUndoToTutorial` (and rename it something like `SendUndoRedoToTutorial`). Also it seems you've accidentally duplicated the auto evo prediction sending in this method.

**Review by hhyyrylainen — CHANGES_REQUESTED** [review]
I started off wanting to write just a few comments but ended up doing a full review. See my previous comments.

File: src/Localization.txt [hunk 9]
```diff
@@ -0,0 +1 @@
+AUTO_RESOLUTION => "auto({0}x{1})"
```
Comment (hhyyrylainen) [inline src/Localization.txt:1]: This file needs to be deleted

File: src/general/OptionsMenu.cs [hunk 10]
```diff
@@ -581,7 +592,7 @@ private void SwitchMode(OptionsMode mode)
                 tutorialsEnabled.Hide();
                 optionsMode = OptionsMode.MainMenu;
                 break;
-            }
+             }
```
Comment (hhyyrylainen) [inline src/general/OptionsMenu.cs:595]: extra space got added here

File: src/general/OptionsMenu.cs [hunk 11]
```diff
@@ -600,6 +611,15 @@ private void SwitchMode(OptionsMode mode)
         }
     }
 
+    // Displays the current Resolution in options.
+    private void DisplayResolution()
+    {
+            var autoResolution = new LocalizedString("AUTO_RESOLUTION",
+                GetViewportRect().Size.x.ToString(CultureInfo.CurrentCulture),
+                GetViewportRect().Size.y.ToString(CultureInfo.CurrentCulture));
+            resolution.Text = autoResolution.ToString();
```
Comment (hhyyrylainen) [inline src/general/OptionsMenu.cs:620]:
```suggestion
            var resolution = GetViewportRect().Size;
            resolution.Text = string.Format(CultureInfo.CurrentCulture, TranslationServer.Translate("AUTO_RESOLUTION"), resolution.x, resolution.y);
```

I think something like that should work, might need slight tweaking for it to compile.

File: src/microbe_stage/editor/MicrobeEditorReportComponent.tscn [hunk 12]
```diff
@@ -755,4 +755,4 @@ __meta__ = {
 [editable path="MarginContainer/HSplitContainer/StatisticsPanel/MarginContainer/ScrollContainer/MarginContainer/VBoxContainer/SpeciesPopulation"]
 [editable path="MarginContainer/HSplitContainer/StatisticsPanel/MarginContainer/ScrollContainer/MarginContainer/VBoxContainer/PhysicalConditions"]
 [editable path="MarginContainer/HSplitContainer/StatisticsPanel/MarginContainer/ScrollContainer/MarginContainer/VBoxContainer/AtmosphericGasses"]
-[editable path="MarginContainer/HSplitContainer/StatisticsPanel/MarginContainer/ScrollContainer/MarginContainer/VBoxContainer/Compounds"]
+[editable path="MarginContainer/HSplitContainer/StatisticsPanel/MarginContainer/ScrollContainer/MarginContainer/VBoxContainer/Compounds"]
```
Comment (hhyyrylainen) [inline src/microbe_stage/editor/MicrobeEditorReportComponent.tscn:758]: Still changes that are not required in unrelated files.

File: locale/en.po [hunk 13]
```diff
@@ -1577,19 +1577,23 @@ msgstr "Mod Loading Errors"
 msgid "MOD_LOAD_ERRORS_OCCURRED"
 msgstr "Errors occurred when loading one or more mods. Logs may contain additional information."
 
-#: ../src/general/OptionsMenu.cs:900
+#: ../src/general/OptionsMenu.cs:621 ../src/general/OptionsMenu.tscn:290
+msgid "AUTO_RESOLUTION"
+msgstr "AUTO({0}x{1})"
```
Comment (hhyyrylainen) [inline locale/en.po:1582]:
```suggestion
msgstr "AUTO ({0}x{1})"
```

File: src/general/OptionsMenu.cs [hunk 14]
```diff
@@ -600,6 +611,19 @@ private void SwitchMode(OptionsMode mode)
...
 
+    /// <summary>
+    /// Displays the current viewport resolution
+    /// </summary>
+    private void DisplayResolution()
+    {
+        if (resolution == null)
+            return;
+
+        Vector2 screenResolution = GetViewportRect().Size;
```
Comment (hhyyrylainen) [inline src/general/OptionsMenu.cs:622]:
```suggestion
        var screenResolution = GetViewportRect().Size;
```

**Review by hhyyrylainen — CHANGES_REQUESTED** [review]
I've now playtested this and the tutorial change works correctly, however the resolution display is not right. I added also one var style comment on it, but it seems that `GetViewportRect` returns the "logical" size of the draw area, not the actual rendering resolution. I can't remember where but I think there was talk or code about how to get the actual rendering resolution from Godot.

File: src/general/OptionsMenu.cs [hunk 15]
```diff
@@ -600,6 +611,18 @@ private void SwitchMode(OptionsMode mode)
...
     }
 
+    /// <summary>
+    /// Displays the current viewport resolution
+    /// </summary>
+    private void DisplayResolution()
+    {
+        if (resolution == null)
+            return;
+        var screenResolution = OS.GetScreenSize();
```
Comment (hhyyrylainen) [inline src/general/OptionsMenu.cs:621]: This is also incorrect. This shows the resolution of the *monitor* the game is on and not the size of the game window / the resolution the game is rendering at.

**Review by hhyyrylainen — CHANGES_REQUESTED** [review]

File: src/general/OptionsMenu.cs [hunk 16]
```diff
@@ -600,6 +611,19 @@ private void SwitchMode(OptionsMode mode)
...
 
+    /// <summary>
+    /// Displays the current viewport resolution
+    /// </summary>
+    private void DisplayResolution()
+    {
+        if (resolution == null)
+            return;
+
+        var screenResolution = OS.WindowSize * OS.GetScreenScale();
```
Comment (hhyyrylainen) [inline src/general/OptionsMenu.cs:622]: This seems to now get the right resolution.

**Review by hhyyrylainen — APPROVED** [review]
Now this seems correct. Just need to fix the merge conflicts and this is good to go. You may want to save the English translation you wrote in a new file and then just revert all the locale files from master, rerun the localization update and copy back that one English translation you made.

**Review by Kasterisk — APPROVED** [review]

**Comment by hhyyrylainen** [comment]
Looks like some of the gettext headers also have gotten out of order with the latest commit.

**Comment by astral-grayson** [comment]
> Someone else earlier today also had the gettext header order problem and their issue was that they had some ancient gettext version they got from somewhere.

That was me! Assuming @harchrisdytian is on Windows, I fixed my setup by installing gettext from [here](https://mlocati.github.io/articles/gettext-iconv-windows.html) as linked to from the GNU site. The current gettext version is 0.21

**Comment by hhyyrylainen** [comment]
The checks now pass but this has a lot of merge conflicts. It's usually easier to just replace all of the .po files with ones from master and rerun the translation update, and then finally put back in the English translations you've done.

**Comment by hhyyrylainen** [comment]
Github still shows this PR as having merge conflicts. I think I'll solve them myself in order to get this ready in time for 0.5.8.1
